### PR TITLE
add custom secret patterns

### PR DESCRIPTION
**Pull request by tim-thacker-nullify**
## Description

add custom secret pattern configuration options

## Checklist

- [ ] I have added one of the `patch`, `minor`, `major` or `no-release` labels
- [ ] I have linked an issue from this repository using the **Development** option
- [ ] I have performed a self-review of my own code
- [ ] I have checked for redundant or commented out code
- [ ] I have commented my code where I can't make it self-documenting
- [ ] I have made corresponding changes to the documentation
- [ ] I have added any appropriate tests

### CODE REVIEW

File: pkg/models/secrets.go [hunk 1]
```diff
@@ -1,7 +1,8 @@
 package models
 
 type Secrets struct {
-	Ignore []SecretsIgnore `yaml:"ignore,omitempty"`
+	Ignore         []SecretsIgnore                 `yaml:"ignore,omitempty"`
+	CustomPatterns map[string]SecretsCustomPattern `yaml:"custom_patterns,omitempty"`
```
Comment (tim-thacker-nullify) [inline pkg/models/secrets.go:5]: @juicetin i set this to a map so we can do a map merge between the global and repo config. i dont think there would be a case where customers want to completely override the whole map with a repo level config.

but very likely there could be a secret pattern a dev wants to put in a specific repo and still use all the global ones.

Comment (juicetin) [reply pkg/models/secrets.go:5]: Yeah I reckon that makes sense to be the default behavior 

If the ignore globals use case starts to come up, we could introduce something like an `ignore_globals` field at each level of the config? _Could_ make for confusion though, would need to think more carefully about the actual use case when it does come up haha

Comment (tim-thacker-nullify) [reply pkg/models/secrets.go]: i added in `custom_patterns_override_global` and implemented the merge logic for it

File: pkg/models/secrets.go [hunk 2]
```diff
@@ -1,7 +1,9 @@
 package models
 
 type Secrets struct {
-	Ignore []SecretsIgnore `yaml:"ignore,omitempty"`
+	Ignore                       []SecretsIgnore                 `yaml:"ignore,omitempty"`
+	CustomPatterns               map[string]SecretsCustomPattern `yaml:"custom_patterns,omitempty"`
+	CustomPatternsOverrideGlobal bool                            `yaml:"custom_patterns_override_global,omitempty"`
```
Comment (juicetin) [inline pkg/models/secrets.go:6]: oh nice, I didn't consider the case in the DM thread of when we want to both extend the global but also override certain ones in the one config file

**Review by juicetin — APPROVED** [review]
not an action for this PR, just thinking out loud

I wonder where would be a good place to give a bunch of examples of config files - there's the public docs with the Jira stuff/and others sit atm

**Pull request merged**